File: bin/extract-comment.rb
#!/usr/bin/env ruby -w

def comment_reject_filter(comment)
  return (
    comment.length <=  20 ||  # comment's length is less than 10
    comment.length >= 200 ||
    !comment.match(/\A@?[-a-zA-Z0-9_ ,.:`'"\/*+()#]++\z/) ||  # comments isn't composed of
    comment.count("-0-9_ ,.: `'\"\/") / comment.length.to_f >= 0.4  # more than 40% symbols
  )
end

def split_sentence(comment)
  comment.split(/(?<=\.)(?<!e\.g\.|i\.e\.|ex\.)\s++/i)
end

def extract_c_comment(code_string)
  comments = code_string.scan(
    %r{(
      # multi-line comment
      /\*  # comment begin
        [^*]*\*++   # general
        (?:
          [^/*]     # special
          [^*]*\*++ # general
        )*
      /  # comment end
      |  # or
      # single line comment
      (?:
        \s++      # extra whitespace
        //[^\n]++ # comment
      )+
    )}x
  ).flatten

  comments = comments
    .map(&:strip)
    .map { |e| e.sub(/^\/\*+/, '') }  # delete "/*"
    .map { |e| e.sub(/\*+\/$/, '') }  # delete "*/"
    .map { |e| e.sub(/^\/\/++/, '') }  # delete "//"
    .map { |e| e.gsub(/\n\s*+((\*|\/\/)[ \t]*+)*/, ' ') }  # delete "\n" or "\n //"
    .map(&:strip)
    .map(&method(:split_sentence))
    .flatten
    .reject(&method(:comment_reject_filter))
end

def extract_sh_comment(code_string)
  comments = code_string.scan(
    %r{(
      # single line comment
      (?:
        \s++      # extra whitespace
        \#[^\n]++ # comment
      )+
    )}x
  ).flatten

  comments = comments
    .map(&:strip)
    .map { |e| e.sub(/^#/, '') }  # delete "#"
    .map { |e| e.gsub(/\n\s*+(\#[ \t]*+)*/, ' ') }  # delete "\n" or "\n #"
    .map(&:strip)
    .map(&method(:split_sentence))
    .flatten
    .reject(&method(:comment_reject_filter))
end

# --- main ---

target_dir = ARGV[0] || 'target-data'
target_files = Dir.glob("#{target_dir}/**/*.*").reject do |path|
  File.directory?(path) || !File.exist?(path)
end

target_files.each do |file|
  code_str = File.read(file, encoding: 'UTF-8')
    .encode("UTF-16BE", "UTF-8", :invalid => :replace, :undef => :replace, :replace => '?')
    .encode("UTF-8")

  comments = []
  if file =~ /\.(c|cpp|h|hpp|java|cs|php|js|ts|go|scala|mm)$/
    comments = extract_c_comment(code_str)
  elsif file =~ /\.(sh|py|rb|coffee|zsh)$/
    comments = extract_sh_comment(code_str)
  else
    next
  end

  next if comments.empty?

  # write result to stdin
  puts "  =====#{file}=====\n"
  puts "#{comments.join("\n")}\n"
end

File: bin/extract-comment.py
#!/usr/bin/env python3

import sys
import os
import glob
import re
from itertools import chain

if sys.version_info < (3, 5):
    raise RuntimeError("must use python 3.5")

target_dir = sys.argv[1]

def count_chars(text, chars):
    count = 0
    for char in chars:
        count += text.count(char)
    return count

def comment_select_filter(comment):
    return (
        len(comment) > 10 and  # comment's length is more than 10
        re.compile(r"^[-a-zA-Z0-9_ ,.:`'\"/*()]+$").match(comment) and  # comments is composed of
        count_chars(comment, "-0123456789_ ,.:`'\"\/") / len(comment) < 0.4  # less than 40% symbols
    )

def split_sentence(comment):
    splitter = re.compile(r"(?<=\.)(?<!e\.g\.|i\.e\.)\s+", re.I)
    return splitter.split(comment)

def flatten(listOfLists):
    "Flatten one level of nesting"
    return chain.from_iterable(listOfLists)

def extract_c_comment(code_string):
    comment_pattern = re.compile(r"""
        # multi-line comment
        /\*  # comment begin
            [^*]*\*+     # general
            (?:
                [^/*]    # special
                [^*]*\*+ # general
            )*
        /  # comment end
        |  # or
        # single line comment
        (?:
            \s+      # extra whitespace
            //[^\n]+ # comment
        )+
    """, re.X)
    comments = comment_pattern.findall(code_string)
    comments = [x.strip() for x in comments]
    comments = [re.sub(r"^/\*+|\*+/$", "", x) for x in comments] # remove "/*" and "*/"
    comments = [re.sub(r"^//+", "", x) for x in comments] # remove "//"
    comments = [re.sub(r"\n\s*((\*|//)[ \t]*)?", " ", x) for x in comments] # remove "\n" or "\n //"
    comments = [x.strip() for x in comments]
    comments = list(map(split_sentence, comments))
    comments = flatten(comments)
    comments = list(filter(comment_select_filter, comments))
    return comments

def extract_sh_comment(code_string):
    comment_pattern = re.compile(r"""
        # single line comment
        (?:
            \s+      # extra whitespace
            \#[^\n]+ # comment
        )+
    """, re.X)
    comments = comment_pattern.findall(code_string)
    comments = [x.strip() for x in comments]
    comments = [re.sub(r"^#+", "", x) for x in comments] # remove "#"
    comments = [re.sub(r"\n\s*(\#[ \t]*)?", " ", x) for x in comments] # remove "\n" or "\n #"
    comments = [x.strip() for x in comments]
    comments = list(map(split_sentence, comments))
    comments = flatten(comments)
    comments = list(filter(comment_select_filter, comments))
    return comments

for filename in glob.iglob('%s/**/*.*' % target_dir, recursive=True):
    if os.path.isdir(filename):
        continue
    if not os.path.isfile(filename):
        continue

    with open(filename, 'r', encoding='utf-8') as f:
        try:
            code_str = f.read()
        except Exception as e:
            pass

        comments = []
        c_lang_ext = re.compile(r"\.(c|cpp|h|hpp|java|cs|php|js|go)$")
        sh_lang_ext = re.compile(r"\.(sh|py|rb|coffee)$")
        if c_lang_ext.search(filename):
            comments = extract_c_comment(code_str)
        elif sh_lang_ext.search(filename):
            comments = extract_sh_comment(code_str)
        else:
            continue

        if len(comments) == 0:
            continue

        # write result to stdin
        print("  =====%s=====" % filename)
        print("%s" % ("\n".join(comments)))
